A draft of an extension defining attribute for a Tenant Identifier

lennygran opened the pull request:
This extension defines an attribute that can be included within a CloudEvent
to describe a tenant identifier for the event produced by an event source in
multi-tenant systems. This attribute allows intermediaries to apply functional
tenant-based logic, such as routing, scheduling or processing priorities, etc.
without a need to understand the event body.

JemDay commented:
I understand the purpose of this but i think it may need a bit more guidance on usage, eg..

- If events flow across multiple provider platforms is the expectation that tenant-id is preserved, overwritten, augmented, etc ?
- Is the assumption that a 'tenant-id' would be added at the point-of-ingress into a provider platform.
- Should Id's be stripped at point of egress (i'm not sure how that would happen).

lennygran commented:
@JemDay I addressed your comment with the commit above. The rationale is the following: because the tenant-id is defined by the event producer, it should stay with the event unchanged, hence the middleware at multiple provider platforms should just propagate it through.

cneijenhuis commented:
My understanding was that the tenant id is part of the `source` attribute. It "identifies the context in which an event happened." To me, the tenant id is an important part of the context.

An example source from the spec is: `/cloudevents/spec/pull/123` - I'd say either only `cloudevents` or `cloudevents/spec` is the tenant id.

An example source from Azure (see https://docs.microsoft.com/en-us/azure/event-grid/cloudevents-schema#cloudevent-schema): 
`/subscriptions/{subscription-id}/resourceGroups/{resource-group}/providers/Microsoft.Storage/storageAccounts/{storage-account}#blobServices/default/containers/{storage-container}/blobs/{new-file}`
Either `storage-account` or `storage-container` are tenant ids.

My hope is that, if the tenant id is part of the standardized source attribute, performing "routing, scheduling or processing priorities" will be supported by more intermediaries than if the tenant id is an extension attribute.

I see where the PR comes from though, as this isn't clearly spelled out in the `source` description. Maybe you could instead open up a PR that clarifies that a tenant id should be part of it?

lennygran commented:
Yes, @cneijenhuis, the lack of semantic clarity on the `source` attribute is contributed to this new extension. I also see a value of clarifying an existing attribute that can carry the same information instead of loading spec with more extensions.

clemensv commented:
"source" is a URI and as such can identify a tenant easily, for instance: 

urn:tenant.myservice.net:context

(using RFC2141 as one way to do that)
